### ISSUE TITLE
How is fine-tuning performed actually?

### ISSUE THREAD

**Issue by stephenllh** [pengzhiliang/MAE-pytorch#14]
From my understanding, the authors pretrained the MAE on ImageNet, and used the encoder for fine-tuning. In their paper, they did not explain how the masking is handled. During the fine-tuning stage, are the images masked? If not, how does it fit the encoder which accepts only a portion of the input?  I wonder how you handle it. Can you kindly show me in your code?

**Comment by pengzhiliang** [comment]
Hello, in the end-to-end fine-tuning stage, the encoder accepts the full set tokens of image without any masked token. In fact, pre-trained stage just provides a better init for fine-tuning. As you can find the difference in my code: 

[pre-train](https://github.com/pengzhiliang/MAE-pytorch/blob/ba35dbcfb8877edd53a6628d3639091337e0c398/modeling_pretrain.py?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L93):     
```python
x = x + self.pos_embed.type_as(x).to(x.device).clone().detach()

B, _, C = x.shape
x_vis = x[~mask].reshape(B, -1, C) # ~mask means visible

for blk in self.blocks:
    x_vis = blk(x_vis)
```
[fine-tune](https://github.com/pengzhiliang/MAE-pytorch/blob/ba35dbcfb8877edd53a6628d3639091337e0c398/modeling_finetune.py?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L264)
```python
if self.pos_embed is not None:
    x = x + self.pos_embed.expand(B, -1, -1).type_as(x).to(x.device).clone().detach()

for blk in self.blocks:
    x = blk(x)
```

Hope this can help you!

**Comment by stephenllh** [comment]
Thanks! I now have one more question: If we see Transformers for NLP, the sequence dimension is fixed to some arbitrary `max_len` hyperparameter. However, that is not the case here, where the number of tokens (sequence length) differs in pretraining and fine-tuning.

**Comment by chokyungjin** [comment]
I have the same question. 
In the case of pre-train, I know that only visible tokens (25% of the original image) in the encoder are input, but when fine-tuning, the model does not use a mask, so if the entire original image is input, isn't it a different size?

**Comment by pengzhiliang** [comment]
Yes, it is actually different. 
The reason is that the network has no requirements in patch dimensions.
`multi head self-attention` only requires the size of` Q`, `K`,` V` needs consistency.
`FFN` only processes on` C` dimensions.
So, the encoder doesn't care about how many patches it accepts.
